internal/pkg/api/handleEnroll.go

```diff
@@ -252,6 +253,85 @@ func (et *EnrollerT) _enroll(
 				Msg("Error when trying to delete old agent with enrollment id")
 			return nil, err
 		}
+		// deleted, so clear the ID so code below knows it needs to be created
+		agent.Id = ""
+	}
+
+	var agentID string
+	if req.Id != nil && *req.Id != "" {
+		agentID = *req.Id
+
+		// check if the agent with this ID already exists
+		var err error
+		agent, err = et._checkAgent(ctx, zlog, agentID)
+		if err != nil {
+			return nil, err
+		}
+
+		if agent.Id != "" {
+			// confirm that this agent has a set replace token
+			// one is required or replacement of this already enrolled and active
+			// agent is not allowed
+			if agent.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled without a replace token set")
+				return nil, ErrAgentNotReplaceable
+			}
+			if req.ReplaceToken == nil || *req.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled; no replace token given during enrollment")
+				return nil, ErrAgentNotReplaceable
+			}
+			same, err := compareHashAndToken(zlog.With().Str("AgentID", agent.Id).Logger(), agent.ReplaceToken, *req.ReplaceToken, et.cfg.PDKDF2)
+			if err != nil {
+				// issue with hash comparison; reason already logged
+				return nil, ErrAgentNotReplaceable
+			}
+			if !same {
+				// not the same, cannot replace
+				// provides no real reason as that would expose too much information
+				zlog.Debug().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled; replace token didn't match")
+				return nil, ErrAgentNotReplaceable
+			}
+
+			// confirm that its on the same policy
+			// it is not supported to have it the same ID enroll into different policies
+			if agent.PolicyID != policyID {
```

Comment:
There is an unlikely race condition where a user assigns the agent to a new policy as this replacement is happening. The Fleet server state isn't guaranteed to be current and we don't have any kind of lock or transaction in the index to help us avoid this type of problem.

Reply:
A user could unenroll the agent as this function is executing to enroll it again with a new API key is another weird situation that could arise.

Reply:
I don't think we really document specific flows in Fleet Server. Those flows I believe are normally document publicly as Elastic Agent documentation. Let me know where you think I should put this information and I can do that.